### PR TITLE
[BEAM-115] Represent a Pipeline via a list of Top-level Transforms

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
The root node is a synthetic transform which does not appear within the
graph, as it never has any components of note. Instead of referring to a
single "root node" in the Pipeline message, refer to the top-level nodes
which do not have an enclosing PTransform.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @kennknowles 

CC: @lukecwik @dhalperi

**Comment by kennknowles** [comment]
Seems reasonable to me. Is there any particular roadblock that prompted it? (I may have missed some dev list discussion or PR commentary)

**Comment by tgroh** [comment]
Converting a Pipeline is close, but the Components (as I've built them) go through the `AppliedPTransform` class to register `PTransforms`, and there's no way to construct one for the root node (at least in Java)

File: sdks/common/runner-api/src/main/proto/beam_runner_api.proto [hunk 1]
```diff
@@ -92,8 +92,8 @@ message Pipeline {
   // this pipeline.
   Components components = 1;
 
-  // (Required) The id of the PTransform that is the root of the pipeline
-  string root_transform_id = 2;
+  // (Required) The ids of all PTransforms that are not contained within another PTransform
+  repeated string top_level_transforms = 2;
```
Comment (lukecwik) [inline sdks/common/runner-api/src/main/proto/beam_runner_api.proto]: root_transform_ids

Comment (tgroh) [inline sdks/common/runner-api/src/main/proto/beam_runner_api.proto]: Done.

**Comment by lukecwik** [comment]
LGTM

**Comment by tgroh** [comment]
retest this please

**Comment by tgroh** [comment]
retest this please

**Comment by robertwb** [comment]
I actually prefer the pipeline having a single root note. Semantically (especially now that we're removing pipeline options) it's cleaner if `Pipeline == PTransform<PBegin, ?>` (and eventually we may even want to reify this in the API, i.e. have `Runner.run(PTransform<PBegin, ?>` and move away from the Pipeline object altogether (which is an odd ducke, e.g. it's not a PValue but has an apply)). This would reduce the manual (and error-prone) scope of pairing each pipeline construction with a `run()`. This is a ways off, but could be reflected in the proto representation right away.)

As for being "empty" it would contain parts and the pipeline-level display_data. If we decide to attach other information to PTransforms (such as resource requirements) that would naturally live there as well.

**Pull request closed without merging**